apps/server/lib/lexical/server/project/progress.ex

```diff
@@ -0,0 +1,252 @@
+            Map.put(state.percentages_by_label, trim(label), %{
+              percentage: 0,
+              total: message,
+              finished: %{}
+            })
+
+          %{state | percentages_by_label: percentages}
+
+        "begin" ->
+          token = System.unique_integer([:positive])
+          progress = ProgressValue.new(title: title(label))
+
+          %{
+            state
+            | token_by_label: Map.put(state.token_by_label, trim(label), token),
+              progress: progress
+          }
+
+        _ ->
+          # TODO: refactor and reduce complexity
+          percentages_by_label = percentages(state, label, message)
+          state = %{state | percentages_by_label: percentages_by_label}
+          progress = report_progress(state, label, message)
+          %{state | progress: progress}
+      end
+    end
+
+    defp percentages(state, "compile", file) do
+      Logger.info("compile state: #{inspect(state)}")
+
+      total = state.percentages_by_label["compile"].total
+      finished = state.percentages_by_label["compile"].finished
+      finished? = Map.has_key?(finished, file)
+
+      if not finished? do
+        finished = Map.put(finished, file, true)
+
+        %{
+          "compile" => %{
+            total: total,
+            finished: finished,
+            percentage: trunc(map_size(finished) * 100 / total)
+          }
+        }
+      else
+        state.percentages_by_label
+      end
+    end
+
+    defp percentages(state, _, _) do
+      state.percentages_by_label
+    end
+
+    defp report_progress(state, label, message) do
+      case trim(label) do
+        "compile" ->
+          compile_report_progress(state, message)
+
+        _ ->
+          ProgressValue.new(message: message)
+      end
+    end
+
+    defp compile_report_progress(state, message) do
+      case state.percentages_by_label["compile"] do
+        nil ->
+          nil
+
+        _ ->
+          ProgressValue.new(
+            message: message,
+            percentage: state.percentages_by_label["compile"].percentage
+          )
+      end
+    end
+
+    def trim(label) do
+      label
+      |> String.trim_trailing(".begin")
+      |> String.trim_trailing(".end")
+      |> String.trim_trailing(".prepare")
+    end
+
+    def kind(label) do
+      kind = String.split(label, ".") |> List.last()
+      if kind in ["prepare", "begin", "end"], do: kind, else: "report"
+    end
+
+    defp title(label) when label in @mix_labels do
+      "mix " <> trim(label)
+    end
+
+    defp title(label) do
+      trim(label)
+    end
+  end
+
+  alias Lexical.Project
+  alias Lexical.Protocol.Notifications.CreateWorkDoneProgress
+  alias Lexical.Protocol.Notifications.Progress, as: LSProgress
+  alias Lexical.Protocol.Types.WorkDone.ProgressBegin
+  alias Lexical.Protocol.Types.WorkDone.Progress.Report, as: ProgressReport
+  alias Lexical.Protocol.Types.WorkDone.ProgressEnd
+  alias Lexical.RemoteControl.Api.Messages
```

Comment:
why the aliases?
Just alias `Notifications` and `Types.WorkDone` and you'll have nicer code with a lower cognitive buden.